R/utilities.R

```diff
@@ -80,3 +80,165 @@ NULL
 #' @export
 #' @importFrom ggplot2 unit
 NULL
+
+
+#' format_fortify
+#'
+#' a unified function to format \code{\link[network]{network}}
+#' or \code{\link[igraph:igraph-package]{igraph}} object.
+#'
+#' @param model an object of class \code{\link[network]{network}}
+#'   or \code{\link[igraph:igraph-package]{igraph}}.
+#' @param nodes a nodes object from a call to fortify.
+#' @param weights the name of an edge attribute to use as edge weights when
+#'   computing the network layout, if the layout supports such weights (see
+#'   'Details').
+#'   Defaults to \code{NULL} (no edge weights).
+#' @param arrow.gap a parameter that will shorten the network edges in order to
+#'   avoid overplotting edge arrows and nodes; defaults to \code{0} when the
+#'   network is undirected (no edge shortening), or to \code{0.025} when the
+#'   network is directed. Small values near \code{0.025} will generally achieve
+#'   good results when the size of the nodes is reasonably small.
+#' @param by a character vector that matches an edge attribute, which will be
+#' @param by a character vector that matches an edge attribute, which will be
+#'   used to generate a data frame that can be plotted with
+#'   \code{\link[ggplot2]{facet_wrap}} or \code{\link[ggplot2]{facet_grid}}. The
+#'   nodes of the network will appear in all facets, at the same coordinates.
+#'   Defaults to \code{NULL} (no faceting).
+#' @param scale whether to (re)scale the layout coordinates. Defaults to
+#'   \code{TRUE}, but should be set to \code{FALSE} if \code{layout} contains
+#'   meaningful spatial coordinates, such as latitude and longitude.
+#' @param stringsAsFactors whether vertex and edge attributes should be
+#'   converted to factors if they are of class \code{character}. Defaults to
+#'   the value of \code{getOption("stringsAsFactors")}, which is \code{TRUE} by
+#'   default: see \code{\link[base]{data.frame}}.
+#' @param .lva a "list vertex attributes" function.
+#' @param .gva a "get vertex attributes" function.
+#' @param .lea a "get edges attributes" function.
+#' @param .gea a "get edges attributes" function.
+#' @param .ael a "as edges list" function.
+#'
+#' @return a \code{\link[base]{data.frame}} object.
+#'
+#' @keywords internal
+#'
+format_fortify <- function(
+  model,
+  nodes = NULL,
+  weights = NULL,
+  arrow.gap = ifelse(network::is.directed(model), 0.025, 0),
+  by = NULL,
+  scale = TRUE,
+  stringsAsFactors = getOption("stringsAsFactors"),
+  .lva = NULL,
+  .gva = NULL,
+  .lea = NULL,
+  .gea = NULL,
+  .ael = NULL
+) {
+  # store coordinates
+  nodes <- data.frame(nodes)
+  colnames(nodes) <- c("x", "y")
+
+  # rescale coordinates
+  if (scale) {
+    nodes$x <- scale_safely(nodes$x)
+    nodes$y <- scale_safely(nodes$y)
+  }
+
+  # import vertex attributes
+  if (length(.lva(model)) > 0) {
+    nodes <- cbind.data.frame(
```

Comment:
Are you sure this method does not recreate issue #54?

Reply:
yes, at least all tests were successful (travis/appveyor and testthat).
Basically the result from the `sapply(..., SIMPLIFY = FALSE)` is a named list, so using `cbind.data.frame` recreates the data.frame.

Reply:
Excellent! Well done for spotting that (the documentation for `cbind` is unhelpful IMO).

Reply:
It is/was the side effect of sapply (sometimes helpful).

Here, I'm using the method cbind.data.frame to ensure a data.frame as the final results, instead of letting R guess and decide.